### DISCUSSION TITLE
From where can I get the static library files `.a` and `.lib` so I can statically link them in my linux and windows projects?

### DISCUSSION THREAD

**Discussion by Haidar0096** [ImageMagick/ImageMagick#5645] [Help] [answered]
Here is the problem:
- I downloaded `ImageMagick-7.1.0-49-Q16-x64-dll.exe` and  installed it on my windows pc.
- copied the `include` and `lib` from the installation folder into my project (C project in visual studio)
- compile and ran my project succesfully
- Now I uninstall imagemagick application from my pc
- run the project, and it doesn't run, it says missing dll for Magick Core

So, I was deceived to think that the `lib` folder I copied is enough, but I asked on discord and they told me these `.lib` files contain only stubs. So the question is, where can I get the real `.lib` files so I link them statically and my project becomes self contained?

**Comment by fmw42** [comment]
I am not a Windows expert, but have you tried using one of the static binary builds at https://imagemagick.org/script/download.php#windows

**Reply by Haidar0096** [reply]
that's exactly where I got the `ImageMagick-7.1.0-49-Q16-x64-dll.exe` from. but as I said above, they don't work.

**Reply by fmw42** [reply]
A Windows expert or one of the IM developers will need to advise.

**Reply by Haidar0096** [reply]
not only windows, but windows and linux, I want to build a cross platform app, and I want to statically link imagemagick on all platforms, so I need the static libraries for linux and windows

**Reply by fmw42** [reply]
I am not sure that there is such a cross platform static library.  But as I said, one of the IM developers will need to advise.

**Comment by snibgo** [comment]
For Windows installations, see https://imagemagick.org/script/download.php#windows

Especially see that "portable" versions do not create or need Windows registry entries, so can be copied to any directory. Non-portable versions can't be easily copied to other directories.

**Reply by Haidar0096** [reply]
> For Windows installations, see https://imagemagick.org/script/download.php#windows
> 
> Especially see that "portable" versions do not create or need Windows registry entries, so can be copied to any directory. Non-portable versions can't be easily copied to other directories.

very nice, I downloaded portable x64 version:

![image](https://user-images.githubusercontent.com/49842031/194967122-8bc6419a-291c-45c2-84ab-91a5c0762be0.png)

hoping that it contains the `.lib` files (static libraries), but all it contain is some `.exe` executables, which I can't use to link to image magick.

Any ideas? 
@snibgo

**Comment by snibgo** [comment]
Sorry, I don't use Visual Studio.

**Comment by Haidar0096** [comment, marked as answer]
I solved it finally, took alot of time :(

So first to clear things out, the `.lib` files I was using contain mock (stub) methods apparently, that's why they were not working, according to the folks on discord. So I needed the "real" static files that contain the actual implementation of the header files.

To achieve this follow exactly the steps in this page: https://github.com/ImageMagick/ImageMagick-Windows

It will tell you how to generate the `lib` files and where to find them. After generation, copy/paste them to your project, whether you are using visual studio (add them to linker inputs) or vscode and cmake then add this to cmake (change the names):
```
#ImageMagick
add_definitions( -DMAGICKCORE_QUANTUM_DEPTH=16 )
add_definitions( -DMAGICKCORE_HDRI_ENABLE=0 )
include_directories(deps/ImageMagick/include)
target_link_directories(demo PRIVATE deps/ImageMagick/lib)
file(GLOB IMAGEMAGICK_LIBS deps/ImageMagick/lib/*.lib)
target_link_libraries(demo 
    "${IMAGEMAGICK_LIBS}"
)
``` 

Note: during building the visual studio solution as required from the page (after you have chosen static library option for example), then you might need to select as a C compiler the legacy MSVC compiler in order to build the solutions successfully.
